pyro/infer/renyi_elbo.py

```diff
@@ -0,0 +1,244 @@
+
+        Computes the ELBO as well as the surrogate ELBO that is used to form the gradient estimator.
+        Performs backward on the latter. Num_particle many samples are used to form the estimators.
+        """
+        elbo_trace = []
+        surrogate_elbo_trace = []
+        tensor_holder = None
+        is_vectorized = self.vectorize_particles and self.num_particles > 1
+
+        # grab a vectorized trace from the generator
+        for model_trace, guide_trace in self._get_traces(model, guide, *args, **kwargs):
+            elbo_particle = 0
+            surrogate_elbo_particle = 0
+
+            # compute elbo and surrogate elbo
+            for name, site in model_trace.nodes.items():
+                if site["type"] == "sample":
+                    if is_vectorized:
+                        log_prob_sum = site["log_prob"].reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+                    elbo_particle = elbo_particle + log_prob_sum.detach()
+                    surrogate_elbo_particle = surrogate_elbo_particle + log_prob_sum
+
+            for name, site in guide_trace.nodes.items():
+                if site["type"] == "sample":
+                    log_prob, score_function_term, entropy_term = site["score_parts"]
+                    if is_vectorized:
+                        log_prob_sum = log_prob.reshape(self.num_particles, -1).sum(-1)
+                    else:
+                        log_prob_sum = site["log_prob_sum"]
+
+                    elbo_particle = elbo_particle - log_prob_sum.detach()
+
+                    if not is_identically_zero(entropy_term):
+                        surrogate_elbo_particle = surrogate_elbo_particle - log_prob_sum
+
+                    if not is_identically_zero(score_function_term):
+                        surrogate_elbo_particle = (surrogate_elbo_particle +
+                                                   (self.alpha / (1. - self.alpha)) * log_prob_sum)
+
+            if is_identically_zero(elbo_particle):
+                if tensor_holder is not None:
+                    elbo_particle = tensor_holder.new_zeros(tensor_holder.shape)
+                    surrogate_elbo_particle = tensor_holder.new_zeros(tensor_holder.shape)
+            else:  # elbo_particle is not None
+                if tensor_holder is None:
+                    tensor_holder = elbo_particle.new_empty(elbo_particle.shape)
+                    # change types of previous `elbo_particle`s
+                    for i in range(len(elbo_trace)):
+                        elbo_trace[i] = tensor_holder.new_zeros(tensor_holder.shape)
+                        surrogate_elbo_trace[i] = tensor_holder.new_zeros(tensor_holder.shape)
+
+            elbo_trace.append(elbo_particle)
+            surrogate_elbo_trace.append(surrogate_elbo_particle)
+
+        if tensor_holder is None:
+            return 0.
+
+        if self.num_particles == 1:
```

Comment:
Related to above comment - can we simplify this and only distinguish between vectorized and non-vectorized case. From what I can see, with `n=1` the weight would be immaterial.

Reply:
But I guess we don't want to operate `log_sum_exp` for the case `n=1`.

Reply:
now, I get what you mean. Will change things accordingly soon. :)